Align `stroke-width` & `stroke-dashoffset` CSS property parsing with the specification

cdumez opened the pull request:
#### 77685f21f584083f4170cb835c0d517448e46912
<pre>
Align `stroke-width` &amp; `stroke-dashoffset` CSS property parsing with the specification
<a href="https://bugs.webkit.org/show_bug.cgi?id=249510">https://bugs.webkit.org/show_bug.cgi?id=249510</a>

Reviewed by Cameron McCormack.

Align `stroke-width` &amp; `stroke-dashoffset` CSS property parsing with the specification:
- <a href="https://svgwg.org/svg2-draft/painting.html#StrokeWidth">https://svgwg.org/svg2-draft/painting.html#StrokeWidth</a>
- <a href="https://svgwg.org/svg2-draft/painting.html#StrokeDashoffsetProperty">https://svgwg.org/svg2-draft/painting.html#StrokeDashoffsetProperty</a>

For stroke-width:
- Use `&lt;length-percentage [0,inf]&gt; | &lt;number [0,inf]&gt;` parser grammar instead of
  `&lt;length-percentage svg&gt;`. The specification actually says
  `&lt;length-percentage&gt; | &lt;number&gt;` but carries on to say that negative values are
  invalid.

For stroke-dashoffset:
- Use `&lt;length-percentage&gt; | &lt;number&gt;` parser grammar instead of
  `&lt;length-percentage svg&gt;`, as per the specification.

* LayoutTests/fast/css/parsing-stroke-width-expected.txt:
* LayoutTests/fast/css/parsing-stroke-width.html:
Update existing test to reflect behavior change.

* LayoutTests/fast/css/stroke-dashoffset-parsing-expected.txt: Added.
* LayoutTests/fast/css/stroke-dashoffset-parsing.html: Added.
* LayoutTests/fast/css/stroke-width-parsing-expected.txt: Added.
* LayoutTests/fast/css/stroke-width-parsing.html: Added.
Extend layout test coverage. Those 2 tests are passing in both Chrome and
Firefox.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-dashoffset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-dashoffset.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-width-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-width.html:
Update existing CSS-Typed-OM tests to match the specification for `stroke-width`
and `stroke-dashoffset`. The tests incorrectly didn&apos;t allow &lt;number&gt; for these
properties.

* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-dashoffset-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-width-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-width-invalid-expected.txt:
Rebaseline WPT tests now that more checks are passing.

* Source/WebCore/css/CSSProperties.json:
- Update parser grammar to match the specification.
- Use convertLengthAllowingNumber() instead of convertLength() since we now allow numbers which need
  to be converted to a length by adding &apos;px&apos; (as per spec).

* Source/WebCore/css/typedom/CSSUnitValue.cpp:
(WebCore::isValueOutOfRangeForProperty):
Update function to correctly report that `stroke-width` doesn&apos;t allow negative values.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertLength):
Introduce a new convertLengthAllowingNumber() function which mostly behaves like
convertLength() but also allows numbers, which it silently converts a lengths
by assuming they are pixels. This behavior is used by `stroke-width` and
`stroke-dashoffset`, and mandated by the specification.

Canonical link: <a href="https://commits.webkit.org/258107@main">https://commits.webkit.org/258107@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ff47882d05af6c09c5478628997b83fc9dbaf45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100932 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33982 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110237 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104920 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/956 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108077 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106715 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34943 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3789 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9901 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44012 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5561 "Built successfully and passed tests") | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5564 "Failed to checkout and rebase branch from PR 7800") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on LayoutTests/fast/css/stroke-dashoffset-parsing.html:
```diff
@@ -0,0 +1,41 @@
+
+// Zero value.
+evalAndLog("target.style.strokeDashoffset = '0';");
+shouldBeTrue("target.style.strokeDashoffset == '0' || target.style.strokeDashoffset == '0px'"); // Gecko says 0px, Blink says 0.
```
Why do we allow either? css-values-3 says:

> However, if a 0 could be parsed as either a &lt;number> or a [&lt;length>](https://drafts.csswg.org/css-values-3/#length-value) in a property (such as [line-height](https://drafts.csswg.org/css2/#propdef-line-height)), it must parse as a &lt;number>.

so the value must be interpreted as a &lt;number>. Unless there's some other spec text somewhere that allows for either serialization.

cdumez replied:
Yes, I hesitated. The reason I allowed both is that WebKit currently doesn't get it right and I didn't fix this yet.
So if I check for 0, we will have a FAIL line in our tests (which I guess is OK).

The issue is that our code currently calls consumeLength() and then falls back to calling consumeNumber() if the earlier didn't work. For 0, consumeLength() currently happily succeeds and returns 0px.

heycam commented on LayoutTests/fast/css/stroke-width-parsing.html:
```diff
@@ -0,0 +1,41 @@
+
+// Zero value.
+evalAndLog("target.style.strokeWidth = '0';");
+shouldBeTrue("target.style.strokeWidth == '0' || target.style.strokeWidth == '0px'"); // Gecko says 0px, Blink says 0.
```
Same here.

weinig commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -4977,8 +4977,8 @@
-                "converter": "Length",
-                "parser-grammar": "<length-percentage svg>"
+                "converter": "LengthAllowingNumber",
+                "parser-grammar": "<length-percentage> | <number>"
```
Since this matches the SVG2 draft definition, perhaps we should update the specification link to that:

https://svgwg.org/svg2-draft/painting.html#StrokeDashoffsetProperty?


(There is also https://drafts.fxtf.org/fill-stroke-3/#stroke-dashoffset, which defines this to `<length-percentage>`)

cdumez replied:
Sure, I can update to the SVG2 spec link. It is unfortunate the the fill-stroke-3 spec is not in agreement though :(
The good thing is that this change is aligning us with Chrome too, not just the SVG2 spec. The bad thing is that it seems Firefox  still matches our old behavior :/

weinig commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -5067,8 +5067,8 @@
-                "converter": "Length",
-                "parser-grammar": "<length-percentage svg>"
+                "converter": "LengthAllowingNumber",
+                "parser-grammar": "<length-percentage [0,inf]> | <number [0,inf]>"
```
Since this matches the SVG2 draft definition, perhaps we should update the specification link to that (instead of https://drafts.fxtf.org/fill-stroke-3/#propdef-stroke-width, which is currently defined to `<length-percentage [0,inf]>#`):

https://svgwg.org/svg2-draft/painting.html#StrokeWidth

cdumez committed:
Align `stroke-width` & `stroke-dashoffset` CSS property parsing with the specification
https://bugs.webkit.org/show_bug.cgi?id=249510

Reviewed by Cameron McCormack.

Align `stroke-width` & `stroke-dashoffset` CSS property parsing with the specification:
- https://svgwg.org/svg2-draft/painting.html#StrokeWidth
- https://svgwg.org/svg2-draft/painting.html#StrokeDashoffsetProperty

For stroke-width:
- Use `<length-percentage [0,inf]> | <number [0,inf]>` parser grammar instead of
  `<length-percentage svg>`. The specification actually says
  `<length-percentage> | <number>` but carries on to say that negative values are
  invalid.

For stroke-dashoffset:
- Use `<length-percentage> | <number>` parser grammar instead of
  `<length-percentage svg>`, as per the specification.

* LayoutTests/fast/css/parsing-stroke-width-expected.txt:
* LayoutTests/fast/css/parsing-stroke-width.html:
Update existing test to reflect behavior change.

* LayoutTests/fast/css/stroke-dashoffset-parsing-expected.txt: Added.
* LayoutTests/fast/css/stroke-dashoffset-parsing.html: Added.
* LayoutTests/fast/css/stroke-width-parsing-expected.txt: Added.
* LayoutTests/fast/css/stroke-width-parsing.html: Added.
Extend layout test coverage. Those 2 tests are passing in both Chrome and
Firefox.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-dashoffset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-dashoffset.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-width-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/stroke-width.html:
Update existing CSS-Typed-OM tests to match the specification for `stroke-width`
and `stroke-dashoffset`. The tests incorrectly didn't allow <number> for these
properties.

* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-dashoffset-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-width-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-width-invalid-expected.txt:
Rebaseline WPT tests now that more checks are passing.

* Source/WebCore/css/CSSProperties.json:
- Update parser grammar to match the specification.
- Use convertLengthAllowingNumber() instead of convertLength() since we now allow numbers which need
  to be converted to a length by adding 'px' (as per spec).

* Source/WebCore/css/typedom/CSSUnitValue.cpp:
(WebCore::isValueOutOfRangeForProperty):
Update function to correctly report that `stroke-width` doesn't allow negative values.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertLength):
Introduce a new convertLengthAllowingNumber() function which mostly behaves like
convertLength() but also allows numbers, which it silently converts a lengths
by assuming they are pixels. This behavior is used by `stroke-width` and
`stroke-dashoffset`, and mandated by the specification.

Canonical link: https://commits.webkit.org/258107@main
